homeassistant/components/media_player/sonos.py

```diff
@@ -714,12 +733,15 @@ def source_list(self):
     @property
     def source(self):
         """Name of the current input source."""
-        if self._is_playing_line_in:
-            return SUPPORT_SOURCE_LINEIN
-        if self._is_playing_tv:
-            return SUPPORT_SOURCE_TV
-
-        return None
+        if self._coordinator:
+            return self._coordinator.source
+        else:
+            if self._is_playing_line_in:
+                return SUPPORT_SOURCE_LINEIN
+            elif self._is_playing_tv:
+                return SUPPORT_SOURCE_TV
+            else:
+                return None
```

Comment:
And if you're able to, obviously tests would be much more preferable with changes like this.